Core/Src/main.c

```diff
@@ -115,9 +122,25 @@ void StartDefaultTask(void *argument);
 /* Private user code ---------------------------------------------------------*/
 /* USER CODE BEGIN 0 */
 
+/* the following reroutes printf to uart */
+#ifdef __GNUC__
+#define PUTCHAR_PROTOTYPE int __io_putchar(int ch)
+#else
+#define PUTCHAR_PROTOTYPE int fputc(int ch, FILE *f)
+#endif
+
+PUTCHAR_PROTOTYPE
+{
+  HAL_UART_Transmit(&huart4, (uint8_t *)&ch, 1, HAL_MAX_DELAY);
+  return ch;
+}
+
 int _write(int file, char* ptr, int len) {
-  HAL_UART_Transmit_DMA(&huart4, (uint8_t *)ptr, len);
-  
+  int DataIdx;
+
+  for (DataIdx = 0; DataIdx < len; DataIdx++) {
+    __io_putchar( *ptr++ );

```

Comment:
messed up diff this stuff was changed on main??

Reply:
I changed this cuz printing was broken and i think it was cuz of the usart dma changes.

Reply:
hmm did u use `\r`?  Also make sure to use `ner serial` or else it wont work.